src/main/java/org/cloudburstmc/proxypass/network/bedrock/session/UpstreamPacketHandler.java

```diff
@@ -147,4 +182,88 @@ private void initializeProxySession() {
         });
     }
 
+    private void initializeOnlineProxySession() {
+        log.debug("Initializing proxy session");
+        this.proxy.newClient(this.proxy.getTargetAddress(), downstream -> {
+            try {
+                if (onlineLoginChain == null) {
+                    initOnlineLoginChain(mcChain);
+                }
+            } catch (Exception e) {
+                log.error("Failed to get login chain", e);
+            }
+
+            downstream.setCodec(ProxyPass.CODEC);
+            downstream.setSendSession(this.session);
+            this.session.setSendSession(downstream);
+
+            ProxyPlayerSession proxySession = new ProxyPlayerSession(
+                this.session, 
+                downstream, 
+                this.proxy, 
+                this.authData, 
+                new KeyPair(mcChain.publicKey(), mcChain.privateKey()));
+            this.player = proxySession;
+
+            downstream.setPlayer(proxySession);
+            this.session.setPlayer(proxySession);
+
+            try {
+                player.getLogger().saveJson("skinData", this.skinData);
+            } catch (Exception e) {
+                log.error("JSON output error: " + e.getMessage(), e);
+            }
+
+            String skinData = ForgeryUtils.forgeSkinData(proxySession.getProxyKeyPair(), this.skinData);
+
+            LoginPacket login = new LoginPacket();
+
+            login.getChain().addAll(onlineLoginChain);
+
+            login.setExtra(skinData);
+            login.setProtocolVersion(ProxyPass.PROTOCOL_VERSION);
+
+            downstream.setPacketHandler(new DownstreamInitialPacketHandler(downstream, proxySession, this.proxy, login));
+            downstream.setLogging(true);
+
+            RequestNetworkSettingsPacket packet = new RequestNetworkSettingsPacket();
+            packet.setProtocolVersion(ProxyPass.PROTOCOL_VERSION);
+            downstream.sendPacketImmediately(packet);
+
+            //SkinUtils.saveSkin(proxySession, this.skinData);
+        });
+    }
+
+    private void initOnlineLoginChain(StepMCChain.MCChain mcChain) throws Exception {
+        String publicKey = Base64.getEncoder().encodeToString(mcChain.publicKey().getEncoded());
+
+        GsonDeserializer<Map<String, ?>> gsonDeserializer = new GsonDeserializer<>(new GsonBuilder().setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE).disableHtmlEscaping().create());
+
+        if (MOJANG_PUBLIC_KEY == null) {
+            initMojangPublicKey();
+        }
+
+        // adapted from https://github.com/RaphiMC/ViaBedrock/blob/main/src/main/java/net/raphimc/viabedrock/protocol/packets/LoginPackets.java#L276-L291
```

Comment:
You should use a permalink here else the line numbers will point into random code when I update that class in the future

Reply:
https://github.com/CloudburstMC/ProxyPass/pull/47/commits/7018fa50bc1c273bbafba2089abfc5f3c542b709